adapters/tlsio_wolfssl.c

```diff
@@ -924,7 +911,10 @@ void tlsio_wolfssl_dowork(CONCRETE_IO_HANDLE tls_io)
         if ((tls_io_instance->tlsio_state != TLSIO_STATE_NOT_OPEN) &&
             (tls_io_instance->tlsio_state != TLSIO_STATE_ERROR))
         {
-            decode_ssl_received_bytes(tls_io_instance);
+            if (tls_io_instance->tlsio_state != TLSIO_STATE_OPENING_UNDERLYING_IO) //Proxy used
```

Comment:
In general flow of state is 

TLSIO_STATE_NOT_OPEN -> TLSIO_STATE_OPENING_UNDERLYING_IO -> TLSIO_STATE_IN_HANDSHAKE -> TLSIO_STATE_OPEN

and when in TLSIO_STATE_OPENING_UNDERLYING_IO we do xio_open (line : 798), and when xio_open completes we move state to TLSIO_STATE_IN_HANDSHAKE. 

Curious to know what is the state transition in proxy, that is causing do work to call in TLSIO_STATE_OPENING_UNDERLYING_IO state.

Reply:
When NOT using a proxy, the calls run `tlsio_wolfssl_open -> xio_open -> socketio_open -> (tlsio_wolfssl.c)on_underlying_io_open_complete`.  Here the tlsio state changes to TLSIO_STATE_IN_HANDSHAKE, followed by the callback `on_handshake_done` which changes the state to TLSIO_STATE_OPEN.  This is what you describe above and is what we are used to.

When using a proxy, the calls run `tlsio_wolfssl_open -> xio_open -> http_proxy_io_open -> xio_open -> socketio_open -> (http_proxy_io.c)on_underlying_io_open_complete -> xio_send ->socketio_send -> (http_proxy_io.c)unchecked_on_send_complete`.  The tlsio state remains in TLSIO_STATE_OPENING_UNDERLYING_IO. 

Following the mbedtls implementation, it is during the do_work section (not the open) that the callback `(tlsio_wolfssl.c)on_underlying_io_open_complete` is called.  `tlsio_wolfssl_dowork -> xio_dowork -> http_proxy_io_dowork -> xio_dowork -> socketio_dowork -> (http_proxy_io.c)on_underlying_io_bytes_received -> (tlsio_wolfssl.c)on_underlying_io_open_complete`.  

IF the state is still in TLSIO_STATE_OPENING_UNDERLYING_IO and the client tries to decode received bytes, the connection fails.  Dowork must occur first.

Reply:
I agree that would be a cleaner solution.  Since it is different than how wolfssl is written, my concern was the possibility of introducing a bug to wolfssl with it, but that would only pertain to the state `TLSIO_STATE_CLOSING`.  Does this state need to run `decode_ssl_received_bytes`?